### PR TITLE
notifcation attribute and event

### PR DESCRIPTION
**Pull request by dasistweb-gmbh**
### 1. Why is this change necessary?
Because nobody was able to take affect to the notification function by valid subscription or custom attributes

### 2. What does this change do, exactly?
Add a Table s_articles_notification_attributes.
Add a Event to take affect on the Notification database subscription
Pass through the attributes for the notifications to the mail tempalate

### 6. Checklist

- [x] I have written tests and verified that they fail without my change - there were tests already
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php [hunk 1]
```diff
@@ -227,13 +227,21 @@ public static function onNotifyConfirmAction(Enlight_Event_EventArgs $args)
...
             $notificationConfirmed = false;
+            $json_data = [];
             if (!empty($getConfirmation['hash'])) {
                 $notificationConfirmed = true;
                 $json_data = unserialize($getConfirmation['data']);
                 Shopware()->Db()->query('DELETE FROM s_core_optin WHERE hash=?', [$action->Request()->sNotificationConfirmation]);
             }
             if ($notificationConfirmed) {
-                $sql = '
+                $eventManager = Shopware()->Container()->get('events');
```
Comment (shyim) [inline engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: Don't use Shopware() global

File: engine/Shopware/Models/Article/Notification.php [hunk 2]
```diff
@@ -58,6 +59,17 @@ class Notification extends LazyFetchModelEntity
      * @ORM\JoinColumn(name="mail", referencedColumnName="email")
      */
     protected $customer;
+
+    /**
+     * INVERSE SIDE
+     *
+     * @var \Shopware\Models\Attribute\ArticleNotification
+     *
+     * @Assert\Valid
```
Comment (shyim) [inline engine/Shopware/Models/Article/Notification.php:68]: I am not sure with this valid. After your Migration is runned, all notification would have no attributes. So this should fail then?

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php [hunk 3]
```diff
@@ -246,12 +254,14 @@ public static function onNotifyConfirmAction(Enlight_Event_EventArgs $args)
...
-                    $json_data['sShopPath'],
-                ]);
+                    Shopware()->Db()->query($sql, [
+                        $json_data['notifyOrdernumber'],
+                        $json_data['sNotificationEmail'],
+                        $json_data['sLanguage'],
+                        $json_data['sShopPath'],
+                    ]);
+                }
+
```
Comment (shyim) [inline engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: Maybe here a Event to get the lastInsertId to insert/update the new attributes or better a filter to provide the values of the attribute data?

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php (hunk 1 above)
Comment (dasistweb-gmbh) [reply engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: this is a static function theres no other possibitly i know

File: engine/Shopware/Models/Article/Notification.php (hunk 2 above)
Comment (dasistweb-gmbh) [reply engine/Shopware/Models/Article/Notification.php:68]: this does not fail and is used like at the other models

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php (hunk 3 above)
Comment (dasistweb-gmbh) [reply engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: therefore i added a event to make the insert action yourself, you could use the modle for example

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php (hunk 1 above)
Comment (shyim) [reply engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: You can remove the static from the method. Listener must not be static  😊

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php (hunk 3 above)
Comment (shyim) [reply engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: Shopware_Notification_Save_Notification allows only to cancel the insert operation. So a plugin must cancel currently the operation and insert s_articles_notification and then into s_articles_notification_attributes.
When multiple plugins doing that on that event that would be uggly 🙈

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php (hunk 1 above)
Comment (dasistweb-gmbh) [reply engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: ok, done :D

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php (hunk 3 above)
Comment (dasistweb-gmbh) [reply engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: i refactored it after your recommendation

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php [hunk 4]
```diff
@@ -318,6 +335,10 @@ public static function onRunCronJob(Shopware_Components_Cron_CronJob $job)
                 continue;
             }
 
+            /** @var Shopware\Bundle\AttributeBundle\Service\DataLoader $attributeLoader */
+            $attributeLoader = Shopware()->Container()->get('shopware_attribute.data_loader');
```
Comment (mitelg) [inline engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: $this->get(

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php [hunk 5]
```diff
@@ -271,14 +287,15 @@ public static function onNotifyConfirmAction(Enlight_Event_EventArgs $args)
      *
      * @param Shopware_Components_Cron_CronJob $job
      */
-    public static function onRunCronJob(Shopware_Components_Cron_CronJob $job)
+    public function onRunCronJob(Shopware_Components_Cron_CronJob $job)
     {
         $modelManager = Shopware()->Container()->get('models');
```
Comment (mitelg) [inline engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: $this->get(

File: engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php [hunk 6]
```diff
@@ -271,14 +287,15 @@ public static function onNotifyConfirmAction(Enlight_Event_EventArgs $args)
...
 
         $conn = Shopware()->Container()->get('dbal_connection');
```
Comment (mitelg) [inline engine/Shopware/Plugins/Default/Frontend/Notification/Bootstrap.php]: $this->get(

**Comment by dasistweb-gmbh** [comment]
@mitelg what do i have to do for scheduling the pr?

**Comment by mitelg** [comment]
@dasistweb-gmbh please be patient :blush:  
normally we meet on a daily base for the GitHub things, but today we skipped the meeting :disappointed:

**Comment by dasistweb-gmbh** [comment]
@mitelg ok, alright :) i only wanted to get sure there's nothing i can do

**Pull request closed without merging**